packages/flutter/lib/src/widgets/app_lifecycle_listener.dart

```diff
@@ -0,0 +1,248 @@
+// Copyright 2014 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'dart:ui';
+import 'package:flutter/foundation.dart';
+
+import 'binding.dart';
+
+/// A callback type that is used by [AppLifecycleListener.onExitRequested] to
+/// ask the application if it wants to cancel application termination or not.
+typedef AppExitRequestCallback = Future<AppExitResponse> Function();
+
+/// A listener that can be used to listen to changes in the application
+/// lifecycle.
+///
+/// To listen for requests for the application to exit, and to decide whether or
+/// not the application should exit when requested, create an
+/// [AppLifecycleListener] and set the [onExitRequested] callback. On macOS, in
+/// order to receive application exit requests originating from the OS, the
+/// `NSPrincipalClass` in your `Info.plist` file must be set to
+/// `FlutterApplication`, or a subclass of `FlutterApplication`.
+///
+/// To listen for changes in the application lifecycle state, define an
+/// [onStateChange] callback. See the [AppLifecycleState] enum for details on
+/// the various states.
+///
+/// {@tool dartpad}
+/// This example shows how an application can listen to changes in the
+/// application state.
+///
+/// ** See code in examples/api/lib/widgets/app_lifecycle_listener/app_lifecycle_listener.0.dart **
+/// {@end-tool}
+///
+/// {@tool dartpad}
+/// This example shows how an application can optionally decide to abort a
+/// request for exiting instead of obeying the request.
+///
+/// ** See code in examples/api/lib/widgets/app_lifecycle_listener/app_lifecycle_listener.1.dart **
+/// {@end-tool}
+///
+/// See also:
+///
+/// * [ServicesBinding.exitApplication] for a function to call that will request
+///   that the application exits.
+/// * [WidgetsBindingObserver.didRequestAppExit] for the handler which this
+///   class uses to receive exit requests.
+/// * [WidgetsBindingObserver.didChangeAppLifecycleState] for the handler which
+///   this class uses to receive lifecycle state changes.
+class AppLifecycleListener with WidgetsBindingObserver, Diagnosticable {
+  /// Creates an [AppLifecycleListener].
+  AppLifecycleListener({
+    WidgetsBinding? binding,
+    this.onResume,
+    this.onInactive,
+    this.onHide,
+    this.onShow,
+    this.onPause,
+    this.onRestart,
+    this.onDetach,
+    this.onExitRequested,
+    this.onStateChange,
+  })  : binding = binding ?? WidgetsBinding.instance,
+        _lifecycleState = (binding ?? WidgetsBinding.instance).lifecycleState ?? AppLifecycleState.detached {
+    this.binding.addObserver(this);
+    onStateChange?.call(_lifecycleState);
```

Comment:
This strikes me as a little odd. The state change may have happened a looong time ago? 

Also, if we do decide to leave this in, why only fire onStateChange and not the specific callback for the state?

Reply:
Okay, I suppose I can leave it up to the caller to initialize their initial state from the binding. I was trying to sync any listener with the current initial state. I figured we couldn't fire the specific callbacks because those only fire on a state transition. I guess I wasn't thinking of that callback as indicating a transition, but of course it is. I'll just remove it.